server/src/main/java/org/elasticsearch/action/search/SearchTaskStatus.java

```diff
@@ -0,0 +1,188 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskInfo;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SearchTaskStatus implements Task.Status {
+
+    public static final String NAME = "search_task_status";
+
+    private final boolean readOnly;
+    private final List<PhaseInfo> phases;
+    private final AtomicReference<PhaseInfo> currentPhase;
+
+    SearchTaskStatus() {
+        this.phases = new CopyOnWriteArrayList<>();
+        this.currentPhase = new AtomicReference<>();
+        this.readOnly = false;
+    }
+
+    public SearchTaskStatus(StreamInput in) throws IOException {
+        this.currentPhase = new AtomicReference<>(in.readOptionalWriteable(PhaseInfo::new));
+        this.phases = in.readList(PhaseInfo::new);
+        this.readOnly = true;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalWriteable(currentPhase.get());
+        out.writeList(phases);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    void phaseStarted(String phase, int expectedOps) {
+        assert readOnly == false;
+        boolean result = currentPhase.compareAndSet(null, new PhaseInfo(phase, expectedOps));
+        assert result : "another phase has not properly completed";
+    }
+
+    void phaseCompleted(String phase) {
+        assert readOnly == false;
+        PhaseInfo previous = currentPhase.getAndSet(null);
+        assert previous.name.equals(phase) : "current phase [" + previous.name +
+            "] is not the phase that's been completed [" + phase + "]";
+        phases.add(previous);
+    }
+
+    //TODO we should probably have a different method for failures, and be careful to count a shard only when all of its copies have failed
```

Comment:
+1, we should process a failure here only if all replicas failed. It should be possible to add an abstract method in `InitialSearchPhase` that is called when there is no more replicas available for a shard ?